### Carregamento dos Dados

In [3]:
import pandas as pd

df = pd.read_csv('../data/creditcard.csv')

In [5]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [7]:
df.shape

(284807, 31)

### Exploração dos Dados

In [11]:
# Filtragem de nulos no dataframe

df.isnull().sum().sum()

np.int64(0)

O resultado nos diz que não há nulos...

In [26]:
# Filtragem de duplicados no dataframe

df[df.duplicated()].shape

(1081, 31)

In [27]:
df[df.duplicated()][df['Class'] == 1].shape

C:\Users\habee\AppData\Local\Temp\ipykernel_8932\3819310256.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df[df.duplicated()][df['Class'] == 1].shape


(19, 31)

Temos 1081 transações duplicadas, sendo que 19 delas são da classe minoritária (fraudulenta). Vamos eliminá-las...

In [28]:
df = df.drop_duplicates()

df[df.duplicated()].shape

(0, 31)

### Transformação dos Dados e SMOTE

Antes de aplicarmos o K-means, precisamos transformar nossos dados para um formato que melhor se adeque ao modelo. Em adicão a isso, aplicamos o SMOTE para que possamos compensar a classe minoritária.

Como os dados estão em um PCA, é interessante analisarmos a normalidade multivariada dos nossos dados. Para isso, vamos utilizar o teste de Henze-Zirkler.